kafka_checked/src/main.rs

```diff
@@ -0,0 +1,242 @@
+// https://gitlab.met.no/oda/oda/-/blob/main/internal/services/oda-kvkafka/consume-checked.go?ref_type=heads
+// https://gitlab.met.no/tjenester/oda/-/blob/dev/base/oda-kvkafka/deployment.yaml?ref_type=heads
+
+use chrono::NaiveDateTime;
+use kafka::consumer::{Consumer, FetchOffset, GroupOffsetStorage};
+use quick_xml::de::from_str;
+use serde::Deserialize;
+use std::env;
+use substring::Substring;
+use thiserror::Error;
+use tokio_postgres::NoTls;
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("postgres returned an error: {0}")]
+    Database(#[from] tokio_postgres::Error),
+    #[error(
+        "no Timeseries found for this data - station {:?}, param {:?}",
+        station,
+        param
+    )]
+    NoTs { station: i32, param: i32 },
+}
+
+#[derive(Debug, Deserialize)]
+/// Represents <KvalobsData>...</KvalobsData>
+struct KvalobsData {
+    station: Vec<Stations>,
+}
+#[derive(Debug, Deserialize)]
+/// Represents <station>...</station>
+struct Stations {
+    #[serde(rename = "@val")]
+    val: i32,
+    typeid: Vec<Typeid>,
+}
+#[derive(Debug, Deserialize)]
+/// Represents <typeid>...</typeid>
+struct Typeid {
+    #[serde(rename = "@val")]
+    val: i32,
+    obstime: Vec<Obstime>,
+}
+#[derive(Debug, Deserialize)]
+/// Represents <obstime>...</obstime>
+struct Obstime {
+    #[serde(rename = "@val")]
+    val: String, // avoiding parsing time at this point...
+    tbtime: Vec<Tbtime>,
+}
+#[derive(Debug, Deserialize)]
+/// Represents <tbtime>...</tbtime>
+struct Tbtime {
+    #[serde(rename = "@val")]
+    val: String, // avoiding parsing time at this point...
+    sensor: Vec<Sensor>,
+    kvtextdata: Option<Vec<Kvtextdata>>,
+}
+#[derive(Debug, Deserialize)]
+/// Represents <sensor>...</sensor>
+struct Sensor {
+    #[serde(rename = "@val")]
+    val: i32,
+    level: Vec<Level>,
+}
+/// Represents <level>...</level>
+#[derive(Debug, Deserialize)]
+struct Level {
+    #[serde(rename = "@val")]
+    val: i32,
+    kvdata: Option<Vec<Kvdata>>,
+}
+/// Represents <kvdata>...</kvdata>
+#[derive(Debug, Deserialize)]
+struct Kvdata {
+    #[serde(rename = "@paramid")]
+    paramid: i32,
+    original: Option<f64>,
+    corrected: Option<f64>,
+    controlinfo: Option<String>,
+    useinfo: Option<String>,
+    cfailed: Option<String>,
```

Comment:
Hmm, ok... but I believe the tag is missing if there is nothing in it? Or is there some output you have seen that indicates otherwise?

Reply:
Ah, crap no you are right, cfailed is still there and empty... <cfailed></cfailed>

Reply:
Can handle this at the SQL level... same as the other options?

Reply:
https://docs.rs/postgres-types/latest/postgres_types/trait.ToSql.html#nullability
I think if the struct derives this then it would convert correctly to null?

Reply:
What that doc means is that Rust `None` will get converted to Postgres `NULL`, it doesn't mean `Some("")` will get converted to `NULL`, that will still be `""`, which I believe is what Manuel was concerned about.

Reply:
Yes, I've been looking into it and came up with this, but it's a bit cumbersome:

```rust
#[derive(Debug, Deserialize)]
struct Kvdata {
    ...
    #[serde(default, deserialize_with = "optional")]
    original: Option<f64>,
    ...
    #[serde(default, deserialize_with = "optional")]
    cfailed: Option<String>,
}

fn optional<'de, D, T>(deserializer: D) -> Result<Option<T>, D::Error>
where
    D: Deserializer<'de>,
    T: std::str::FromStr,
    <T as std::str::FromStr>::Err: std::fmt::Debug,
{
    Option::deserialize(deserializer).map(|opt| match opt {
        Some("") => None,
        Some(val) => Some(val.parse::<T>().unwrap()),
        None => None,
    })
}
```

It deserializes to `None` if the field is either missing or empty.

Reply:
Have taken this suggestion in the code.